DARIAEngineering/dcaf_case_management
# Set up acts_as_tenant

I rule and have completed some work on Case Manager that's ready for review!

This sets up the acts_as_tenant gem, to enable us to do multitenancy.

The general approach here is to move a lot of the stuff that's in env vars now to a `Fund` object, and read from that rather than env vars. The rest of it is, frankly, mostly grepping around for `FUND` and removing that in favor of some form of `current_tenant.name`. (Exceptions to this are lines - more on lines later - and fund-mailer-domain, which has been daria-services for a little while now.) The diff is large, but most of it turned out to be find and replace.

Deployment-wise, what this will mean in the immediate term is that every fund's instance will get turned into multitenant apps with only one tenant. We'll need to declare a brief downtime to reconfigure, but the two step plan is basically:

* Create a Fund object with the proper subdomain set
* Run some sql to set every row in every table's fund_id to 1

This approach means we can punt on lines for now, but most likely we'll need to retool to make Line a proper object. (Having line be set on the `Fund` level doesn't work super well because of the enums and validation on other objects right now.) But we can roll this out as is without a problem.

To test:
* try the migration strategy: 

```ruby
git checkout main && rails db:drop db:create db:migrate db:seed && git checkout tenants-2 && rails db:migrate
echo "Fund.create name: 'CatFund', domain: 'supercat', subdomain: 'fluffy', full_name: 'Catte Fund" | rails c
sql="
begin;
update archived_patients set fund_id = 1;
update call_list_entries set fund_id = 1;
update calls set fund_id = 1;
update clinics set fund_id = 1;
update configs set fund_id = 1;
update events set fund_id = 1;
update external_pledges set fund_id = 1;
update fulfillments set fund_id = 1;
update notes set fund_id = 1;
update patients set fund_id = 1;
update practical_supports set fund_id = 1;
update users set fund_id = 1;
update versions set fund_id = 1;
commit;
"
echo $sql | rails db
```

* spin up the app as normal - `git checkout tenants-2 && rails db:drop db:create db:migrate db:seed` then `rails s` and try logging in as normal - localhost:3000 - to confirm you can hit the first fund
* log in to the alt development tenant - http://catbox.lvh.me:3000 to confirm you can hit the second fund

This pull request makes the following changes:
* sets up acts as tenant
* adjusts a lot of stuff to be tenant

no significant view changes

It relates to the following issue #s: 
* bumps #1817 

For reviewer:
* Adjust the title to explain what it does for the notification email to the listserv.
* Tag this PR:
  * `feature` if it contains a feature, fix, or similar. This is anything that contains a user-facing fix in some way, such as frontend changes, alterations to backend behavior, or bug fixes.
  * `dependencies` if it contains library upgrades or similar. This is anything that upgrades any dependency, such as a Gemfile update or npm package upgrade.
* If it contains neither, no need to tag this PR.
